src/net/rpc.rs

```diff
@@ -1333,6 +1335,54 @@ impl ConversationHttp {
         response.send(http, fd).map(|_| ())
     }
 
+    /// Handle a GET to fetch whether or not a contract implements a certain trait
+    fn handle_get_is_trait_implemented<W: Write>(
+        http: &mut StacksHttp,
+        fd: &mut W,
+        req: &HttpRequestType,
+        sortdb: &SortitionDB,
+        chainstate: &mut StacksChainState,
+        tip: &StacksBlockId,
+        contract_addr: &StacksAddress,
+        contract_name: &ContractName,
+        trait_id: &TraitIdentifier,
+    ) -> Result<(), net_error> {
+        let response_metadata = HttpResponseMetadata::from(req);
+        let contract_identifier =
+            QualifiedContractIdentifier::new(contract_addr.clone().into(), contract_name.clone());
+
+        let response =
+            match chainstate.maybe_read_only_clarity_tx(&sortdb.index_conn(), tip, |clarity_tx| {
+                clarity_tx.with_clarity_db_readonly(|db| {
+                    let mut analysis = db.load_contract_analysis(&contract_identifier)?;
+                    if analysis.implemented_traits.contains(trait_id) {
+                        Some(GetIsTraitImplementedResponse {
+                            is_implemented: true,
+                        })
+                    } else {
+                        let trait_definition = analysis.get_defined_trait(&trait_id.name)?;
+                        let is_implemented = analysis
+                            .check_trait_compliance(trait_id, trait_definition)
```

Comment:
Maybe this is a question for @kantai -- how expensive can this get?  Should this API call run with a given `ExecutionCost`?  I could imagine an attacker crafting an evil trait definition and contract that were specifically constructed to force this method to take a lot of CPU time.

Reply:
The trait compliance check is linear with respect to the size of the contract and the trait definition (and much faster than fetching either). This whole evaluation isn't exactly a "free" operation, but it's about the same as the account balance lookup API (it's 2 MARF reads: one for the trait definition and one for the contract in question), plus the read length of the two contracts (absolute maximum is two Stacks block sized reads). I'm not too worried about this becoming a too-expensive endpoint for that reason, it's actually doing much less work than the read-only call interface (which needs to do the contract fetch lookup before it begins any processing). If we wanted to add a bound on this, it isn't something it'd be obvious to add an ExecutionCost barrier to: during analysis, the data fetches are charged after fetch (because you cannot know how large the trait definition contract is until you've consulted the analysis db), so I think we'd need to place just a limit on the contract size that could be evaluated in this endpoint. But again, I don't think it's worthwhile -- an attack on this endpoint would need to make lots and lots of requests, and that's something that would need a proxy to protect against anyways.

Reply:
Ah, okay.  Thanks for clarifying!  Sounds good to me.